### ISSUE TITLE
Overly specific implicit Windows `os` package

### ISSUE THREAD

**Issue by mottosso** [nerdvegas/rez#610]
### Problem

Windows gone did an update overnight, and suddenly all of my packages stopped resolving, saying:

```bash
$ rez env python
The context failed to resolve:
The following package conflicts occurred: (os-windows-10.0.17134 <--!--> ~os==windows-10.0.17763)
```

And after staring myself blind at this seemingly contradictory line, I spotted the last 3 digits have increased since yesterday.

Windows isn't good for much, but it's really good at backwards compatibility. It would be safe to strip those last 3 digits I think. Heck, it'd be safe to strip almost every digit, as anything working on Windows XP is likely still compatible with Windows 10.

On Linux, the implicit packages are specific enough I think.

```bash
~platform==linux  (implicit)
~arch==x86_64     (implicit)
~os==Debian-9.7   (implicit)
```

Here's what they look like on Windows.

```bash
requested packages:
~platform==windows       (implicit)
~arch==AMD64             (implicit)
~os==windows-10.0.17763  (implicit)
```

### Solution

Would recommend making this level of specificity opt-in, to avoid newcomers from falling into this trap. How about a default of 2 digits, e.g. `windows-10.0`?

**Comment by instinct-vfx** [comment]
First: I do agree that the os is overly specific. That being said i usually work around this by not requiring os and not putting it into the variants. I am usually sticking to platform and arch only. The implicit packages are weak references so they are only matched IF the package is required by your package. So that conflict should not arise unless your python package has a very specific os requirement, no? Is this python package created using rez-bind?

Depending on your requirements you could also set a platform map in your config: https://github.com/nerdvegas/rez/wiki/Configuring-Rez#platform_map

There are quite some differences for Win7->Win10 btw. Nothing major but i have not really gotten the hang of their new "versioning" :(

**Comment by smaragden** [comment]
We are using a platform_map in rezconfig t osolve this exact issue.
This is for CentOS but you can do the same for windows
```python
platform_map = {
    "os":{
        r'CentOS-(\d+)\.(\d+)(\.(\d+))?': r'CentOS-\1',
    }
}
```
This will change **CentOS-7.6.1810** to **CentOS-7**

**Comment by mottosso** [comment]
> That being said i usually work around this by not requiring os and not putting it into the variants

Yes, hmm. I hadn't included `os` as a requirement to anything, and was stumped to find it being a requirement all of a sudden, and now I'm having trouble reproducing it. These were hand-made packages without rez-bind, and the only ones present on the `rez_package_path`.

Thanks for sharing your insights, I shall refer back to this issue if and when it re-surfaces. Until then, it still seems worth lowering the specificity per default I think, and facilitate an opt-in for more.

**Comment by bpabel** [comment]
It's been a while since I've used it, but I remember `rez-pip` also including the very specific `os` version in the variant it generates.

**Comment by mottosso** [comment]
>I remember rez-pip also including the very specific os version in the variant it generates.

I can confirm that it is happening. Can also confirm that all of my pip packages no longer work because of it. :(

I would expect:

1. Any pure-Python pip install not to come with a dependency on os, platform or arch
1. Any pip install with any such dependency to only depend on what's specified in the `setup.py`, e.g. Python 3, Windows if it's a binary etc.
1. If binary, a less specific OS version

In practice and ideal circumstances, all machines in a facility would have an identical version - being versioned by e.g. Ansible or Puppet. This issue is more likely to affect smaller shops I think and perhaps more importantly beginners and anyone evaluating Rez.

**Comment by instinct-vfx** [comment]
_Disclaimer: This reply is a bit off-topic for the original issue, but on-topic for your last post. We may want to move discussion on rez-pip specifically to either the mailing list or its own issue._

This is quite an issue for us (current rez-pip variant handling). I have addressed this in my former internal implementation. We plan to also add this to our PR for rez-pip. My implementation detects packages containing binary components by checking file extensions. If binary files are detected then platform, arch and python major.minor is added as variant. If not, then only python is added.

This still leaves issues with pure python compatibility issues (most notably 2.x vs. 3.x but there are also more subtle issues).

The main issue is that meta-data is completely unreliable. There is no means whatsoever that enforces complete, let alone correct meta-data. Depending on the package maintainer your mileage may vary a LOT. To make things worse, there are a LOT of different meta-data versions (spread over many different PEPs) that either contain specific fields or not etc. etc. To add a bit more confusion scripts created through setuptools are handled very differently (on windows they don't use a shebang but an actual compiled binary wrapper meh).

Your ideal case misses a big part of the equation sadly and will not really happen i think (independent of shop size). There are different factors causing this:

- You may simply WANT to support different python versions (and with Rez, that's perfectly valid)
- You may have scripts and libraries that require different python versions (e.g. embedded interpreters in DCCs and multiple of these if you support multiple DCC versions, standalone etc.)
- You may have to support different OSes (Windows for Max, Linux for Houdini, A different distro of linux for webservices etc. etc.)

TL;DR; Python packaging is still a hellhole of a mess and most straight forward approaches are likely to fail. My main line of thought would be::

- Provide an implementation that aims to handle most cases well with sensible defaults
- "Most cases" should include support for Linux, Windows MacOS
- "Most cases" should also either handle rather rare cases (like environment specific dependencies or optional dependencies) or refuse to create a package and force the user to give specific input on how to handle a given package
- Provide ways to override default (=automatically detected) behavior (e.g. add optional parameters to rez-pip to allow enforcing a certain level of dependency/variant "specific-ness")

We aim to do most of this with our WIP PR for rez-pip.

**Comment by bpabel** [comment]
@instinct-vfx The wheel format includes pretty much all of the possible variant information -- https://www.python.org/dev/peps/pep-0427/#file-name-convention

It's probably not even worth trying to support any older metadata format, since everything is standardized around wheels now.

**Comment by instinct-vfx** [comment]
@bpabel Yeah, we try to rely on wheels wherever possible (newer pip versions seem to allow to re-build as wheel and hence standardize). But still, the contents of the metadata field are up to the maintainer. And while these days there's a new "standard" for dependencies a lot of packages still only have them in run_requires.

**Comment by mottosso** [comment]
Even without wheels, I'm curious what metadata you worry about? If e.g. the Python version was wrong, it would affect anyone on pip as well. E.g. `pip3 install` would pick up the wrong variant. On any binary install, the OS is a given, and the arch too.

**Comment by instinct-vfx** [comment]
I was referring to this statement:

`Any pip install with any such dependency to only depend on what's specified in the setup.py, e.g. Python 3, Windows if it's a binary etc.`

I am mostly talking about dependencies here as extracting them is non-trivial. For python, platform and arch (i would advise against os for the very reason of this issue) i agree that it should be based on the environment your installing from with a binary extension.

You can't necessarily rely on anything in the meta-data. And you don't always have access to setup.py (and even if you do, i would advise to use an install process (either wheeling and installing or installing from source) and not parse setup.py directly.
